File: backend/schemas/ipo.py
from pydantic import BaseModel
from typing import Optional, List, Dict


class IPOListItem(BaseModel):
    id: str
    company_name: str
    symbol: Optional[str] = None
    status: str  # "active" | "upcoming" | "listed"
    open_date: str
    close_date: str
    listing_date: Optional[str] = None
    issue_price_min: float
    issue_price_max: float
    lot_size: int
    issue_size_cr: float
    gmp: float
    gmp_pct: float
    subscription_total: float
    subscription_qib: float
    subscription_nii: float
    subscription_rii: float
    recommendation: str       # "INVEST" | "AVOID" | "NEUTRAL"
    recommendation_score: float
    expected_listing_gain_pct: Optional[float] = None
    risk_level: str           # "HIGH" | "MEDIUM" | "LOW"
    sector: str
    registrar: str


class IPOFinancials(BaseModel):
    revenue_cr: Optional[float] = None
    ebitda_cr: Optional[float] = None
    pat_cr: Optional[float] = None
    eps: Optional[float] = None
    pe_ratio: Optional[float] = None
    roe: Optional[float] = None
    debt_equity: Optional[float] = None


class IPODetail(IPOListItem):
    company_description: str = ""
    financials: IPOFinancials = IPOFinancials()
    promoter_holding_pre: Optional[float] = None
    promoter_holding_post: Optional[float] = None
    industry_pe: Optional[float] = None
    objects_of_issue: List[str] = []
    pros: List[str] = []
    cons: List[str] = []
    analysis_summary: str = ""
    lead_managers: List[str] = []
    dimension_scores: Dict[str, float] = {}

File: frontend/src/types/ipo.ts
export interface IPOListItem {
  id: string;
  company_name: string;
  symbol?: string;
  status: 'active' | 'upcoming' | 'listed';
  open_date: string;
  close_date: string;
  listing_date?: string;
  issue_price_min: number;
  issue_price_max: number;
  lot_size: number;
  issue_size_cr: number;
  gmp: number;
  gmp_pct: number;
  subscription_total: number;
  subscription_qib: number;
  subscription_nii: number;
  subscription_rii: number;
  recommendation: 'INVEST' | 'AVOID' | 'NEUTRAL';
  recommendation_score: number;
  expected_listing_gain_pct?: number;
  risk_level: 'HIGH' | 'MEDIUM' | 'LOW';
  sector: string;
  registrar: string;
}

export interface IPOFinancials {
  revenue_cr?: number;
  ebitda_cr?: number;
  pat_cr?: number;
  eps?: number;
  pe_ratio?: number;
  roe?: number;
  debt_equity?: number;
}

export interface IPODetail extends IPOListItem {
  company_description: string;
  financials: IPOFinancials;
  promoter_holding_pre?: number;
  promoter_holding_post?: number;
  industry_pe?: number;
  objects_of_issue: string[];
  pros: string[];
  cons: string[];
  analysis_summary: string;
  lead_managers: string[];
  dimension_scores: Record<string, number>;
}
